### ISSUE TITLE
sc_consensus_slots::check_equivocation is not guaranteed to catch equivocation

### ISSUE THREAD

**Issue by nazar-pc** [paritytech/polkadot-sdk#1302]
Right now `sc_consensus_slots::check_equivocation` relies on `AusStore` to persist headers for a particular slot. However, there is no locking in there that would ensure two concurrently verified headers are checked against each other, moreover, one of them or both can be overridden in case a third header for the same slot is verified concurrently and will never be equivocated.

It seems that either `sc_consensus_slots::check_equivocation` or its caller (both Aura and Babe right now) need to ensure there is some locking present that prevents concurrency when checking equivocation.

**Comment by bkchr** [comment]
Yeah, good catch! Currently we don't use this in parallel, but we should take care it works also in parallel.

**Comment by nazar-pc** [comment]
Isn't verification done concurrently right now? If not then it is also something I'd like to see implemented and will likely get to in near to mid-term future.

**Comment by bkchr** [comment]
> Isn't verification done concurrently right now?

No. Normally there are also not that many different blocks being produced at the same height.

**Comment by nazar-pc** [comment]
I'm mostly curious about sync case where thousands of blocks can be queued and would be great to do verification on them in parallel.

**Comment by bkchr** [comment]
Currently we don't call `verify` before we download the blocks, which we actually should do.

**Comment by nazar-pc** [comment]
Before or after? I'm confused.

**Comment by bkchr** [comment]
Someone announces a block, we don't have the block, so we download it (ignoring some things here). Then we pass the downloaded block to the `ImportQueue` and this first verifies and then imports the block.

**Comment by bkchr** [comment]
> That makes sense, but I don't see how would you verify it when having just a hash.

Block announce is not just the hash, it is the entire header of the block.



> If there is more than 1 block queued I'd like to make verification to happen in parallel for them.

Yeah that makes sense.

**Comment by nazar-pc** [comment]
Now it makes more sense, I was assuming announcement has just a hash.

I see `BlockImportParams.body` is `Option<Vec<Block::Extrinsic>>`. My understanding that depending on sync mode body may or may not be present. Verifying before downloading is interesting, but that is likely a breaking change for verifier that expects the same contents as block import later on. I don't think any consensus implementation in Substrate relies on block verifier to have access to block body, but it still seems like a bit of an issue. Maybe just worth mentioning in the changelog when this modification is done.

Not the highest on my priority list, but I'll work on these stuff a bit later, thanks for clarifying!

**Comment by bkchr** [comment]
Yeah the verifier currently also works on the state of the previous block to do things like verifying the inherents. This should not be done there. Yeah, when we change this, it will be a bigger change that just changing the underlying usage of `Verifier`.
